### PR TITLE
nixos/doc: document fakeNss, binSh

### PR DESCRIPTION
**Pull request by flokli**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: doc/builders/images/dockertools.section.md [hunk 1]
```diff
@@ -312,3 +312,31 @@ buildImage {
 ```
 
 Creating base files like `/etc/passwd` or `/etc/login.defs` is necessary for shadow-utils to manipulate users and groups.
+
+## fakeNss {#ssec-pkgs-dockerTools-fakeNss}
+
+If your primary goal is providing a basic skeleton for user lookups to work,
+and/or a lesser privileged user, adding `pkgs.fakeNss` to
+`build*Image.contents` might be the better choice than a custom script running
```
Comment (roberth) [inline doc/builders/images/dockertools.section.md:320]: I'm not convinced that it's better. The supposedly declarative merging of `contents` is a significant source of uncertainty with `dockerTools`. In fact, merging directories and files is non-trivial, so we ought to provide a well-specified and well-documented method for doing so. Shell commands provide that, even if it's hard to admit that `contents` was a failed design.

File: doc/builders/images/dockertools.section.md [hunk 2]
```diff
@@ -312,3 +312,31 @@ buildImage {
...
+
+It provides a `/etc/passwd` and `/etc/group`, containing `root` and `nobody`
+users and groups.
+
+It also provides a `/etc/nsswitch.conf`, configuring NSS host resolution to
+first check `/etc/hosts`, before checking DNS, as the default in the absence of
+a config file (`dns [!UNAVAIL=return] files`) is quite unexpected.
+
+You usually might to pair it with binSh, which provides `bin/sh` as a symlink
+to `bashInteractive` (as `/bin/sh` is configured as a shell).
```
Comment (roberth) [inline doc/builders/images/dockertools.section.md:331]: I'd expect this to be in a new section for discoverability and focus.

The sentence is a bit broken. You probably meant:

might want to pair

... but I think that's unnecessarily fluffy. How about

```suggestion
You can pair it with `binSh`, which provides `bin/sh` as a symlink
to `bashInteractive` (as `/bin/sh` is configured as a shell).
```

**Review by roberth — CHANGES_REQUESTED** [review]

**Comment by roberth** [comment]
We should reuse NixOS modules instead of `buildEnv`/`symlinkJoin` or other not-quite-fitting solutions. I wish I could spend time on that. It's just a couple of small refactoring PRs away.

File: doc/builders/images/dockertools.section.md (hunk 1 above)
Comment (flokli) [reply doc/builders/images/dockertools.section.md:320]: in the case of `etc/passwd` already being present in another layer, I'd expect the merging (with default priority) to just fail (?).

Shell commands are way less declarative. The code in `examples.nix` before f7ee2706c2aaef359dd05b6910dcbfa6b3493dee, and the `useradd --system --gid nginx nginx` in there wasn't very reproducible - you're not exactly sure what user id `useradd` will pick.

**Comment by flokli** [comment]
> We should reuse NixOS modules instead of `buildEnv`/`symlinkJoin` or other not-quite-fitting solutions. I wish I could spend time on that. It's just a couple of small refactoring PRs away.

I'm not sure I agree - when it comes to evaluation times, the current approach here might be faster. In any case, that's independent of this documentation PR.

File: doc/builders/images/dockertools.section.md [hunk 3]
```diff
@@ -312,3 +312,31 @@ buildImage {
...
+
+```nix
+buildImage {
+  name = "shadow-basic";
+
+  contents = [
+    binSh
+    fakeNss
+  ]
```
Comment (roberth) [inline doc/builders/images/dockertools.section.md:340]: I've deprecated `buildImage` `contents`. You can use `copyToRoot = buildEnv ...` here.
See https://github.com/NixOS/nixpkgs/pull/179801/files

**Comment by roberth** [comment]
> I'm not sure I agree - when it comes to evaluation times, the current approach here might be faster.

The module system is fast. What's slowing down NixOS is the vast number of service modules, which we won't be using in docker images. Those will only use a handful of key modules.

> @roberth what do we do with this?

Sorry for taking this long. It dropped off my radar.

See comment to resolve my main issue. Can merge after that.

**Comment by flokli** [comment]
@roberth I rebased and updated this. PTAL.

**Pull request merged**